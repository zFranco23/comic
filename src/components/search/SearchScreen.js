import React, { useMemo } from 'react'
import queryString from 'query-string';
import { useLocation } from 'react-router-dom';
import { useForm } from '../../hooks/useForm';
import HeroCard from '../heroes/HeroCard';
import { getHeroesByName } from '../../selectors/getHeroesByName';

const SearchScreen = ( { history }) => {
    
    const location = useLocation();
    
    //parseamos los query params 
    const { q= "" } = queryString.parse(location.search); //solo trabajar con el q 
    //si no hay params es undefined , por eso definimos por defecto : ""
    
    const initialForm = {
        searchText: q,
    };
    console.log("ga")
    const [ formValues, handleInputChange] = useForm( initialForm );
    
    const { searchText }  = formValues;

    const heroesFiltered = useMemo(() => getHeroesByName(q) , [q]);

    const handleSearch = (e) => {
        e.preventDefault();
        history.push(`?q=${searchText}`);
    }


    return (
        <div>
            <h1>SearchScreen</h1>
            <hr />

            <div className="row">
                <div className="col-5">
                    <h4>Search Form</h4>
                    <hr />
                    <form onSubmit={handleSearch}>
                        <input 
                            type="text"
                            placeholder="Find your hero"
                            value={searchText}
                            onChange={handleInputChange}
                            name="searchText"
                            className="form-control"
                            autoComplete="off"
                        />
                        
                        <button
                            type="submit"
                            className="btn mt-1 btn-block btn-outline-primary"
                        >
                            Search...
                        </button>
                    </form>
                </div>
                <div className="col-7">

                    <h4>Results</h4>
                    <hr />
                    {
                        q==="" 
                            && 
                            <div className="alert alert-info">
                                Search a hero ...
                            </div>
                    }
                    {
                        q!=="" && heroesFiltered.length===0 &&
                        <div className="alert alert-danger">
                            There is not a hero with {q} ...
                        </div>
                    }
                    {
                        heroesFiltered.map((hero,i)=>(
                            <HeroCard key={i} {...hero} />
                        ))
                    }
                </div>
            </div>
        </div>
    )
}

export default SearchScreen
