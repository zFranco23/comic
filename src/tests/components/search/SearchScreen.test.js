import { mount } from "enzyme"
import { MemoryRouter, Route } from "react-router-dom"
import SearchScreen from "../../../components/search/SearchScreen"




describe('Pruebas en <SearchScreen />', () => {
    
    const historyMock = {
        push : jest.fn(),
    }
    
    test('Debe mostrarse correctamente', () => {
        
        const wrapper = mount(
            <MemoryRouter>
                <SearchScreen history={historyMock}/>
            </MemoryRouter>
        )
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.alert-info').text().trim()).toBe("Search a hero ...");
    })
    
    test('Debe de mostrar a robin y el input con el valor del queryString ', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={["/search?q=robin"]}>
                <Route path="/search">
                    <SearchScreen history={historyMock}/>
                </Route>
            </MemoryRouter>
        )

        expect(wrapper.find('input').prop('value')).toBe('robin');
        expect(wrapper.find('HeroCard').exists()).toBe(true);
    })
  
    
    test('Debe de mostrar un error si no se encuentra el Hero', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={["/search?q=robin123"]}>
                <Route path="/search">
                    <SearchScreen history={historyMock}/>
                </Route>
            </MemoryRouter>
        )

        expect(wrapper.find('.alert-danger').text().trim()).toBe('There is not a hero with robin123 ...');
        expect(wrapper.find('HeroCard').exists()).toBe(false);
    })
    
    test('Debe de llamar el push del history', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={["/search"]}>
                <Route path="/search">
                    <SearchScreen history={historyMock}/>
                </Route>
            </MemoryRouter>
        )

        wrapper.find('input').simulate('change',{
            target:{
                name :"searchText",
                value:"robin"
            }
        })
        wrapper.find('form').prop('onSubmit')({
            preventDefault : () => {}
        });

        expect(historyMock.push).toHaveBeenCalledWith('?q=robin');
    })
    
})
