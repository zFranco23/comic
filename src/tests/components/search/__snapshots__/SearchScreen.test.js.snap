// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchScreen /> Debe mostrarse correctamente 1`] = `
<div>
  <h1>
    SearchScreen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value=""
        />
        <button
          className="btn mt-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search a hero ...
      </div>
    </div>
  </div>
</div>
`;
